@tayrinn Kim Olga

Tayrinn opened the pull request:
У меня нет планшета, а на ноутбуке не работают эмуляторы из-за недостатка оперативки =(
Поэтому сделала в случае портретной ориентации - viewpager и второй фрагемнт открывается во весь экран. А когда альбомная ориентация - мастер-детейл и диалог-фрагмент. При перевороте экрана состояния сохраняются и все списки проматываются.
Кучу времени потратила чтобы прикрутить красивости, но так и не вышло( sharedTransitions меня победили. Отрисовка фрагмента с табами ужасно тормозит, а как поправить это я так и не нашла.

Tayrinn committed:
add viewpager and retain fragment

Tayrinn committed:
add listfragment, dialogfragment

Tayrinn committed:
fix orientation change bug

zagayevskiy commented:
Понятно, что ты не смогла проверить, но надо было хотя бы попробовать использовать sw600dp и т.д. Типа, теоретически оно бы завелось. А так - неплохо, даже пейджер не моргает при возврате из деталей, это плюс.

zagayevskiy commented on app/src/main/java/ru/yandex/yamblz/ui/activities/MainActivity.java:
```diff
@@ -26,11 +54,124 @@ protected void onCreate(@Nullable Bundle savedInstanceState) {
+
+    private void updateArtistsInFragmentsByTag(String tag, List<Artist> artists) {
+        Fragment fragment = getSupportFragmentManager().findFragmentByTag(tag);
+        if (fragment != null && fragment instanceof UpdateArtistsListener) {
```
Можно сократить до fragment instanceof UpdateArtistsListener, потому что null не экземпляр любого класса.

zagayevskiy commented on app/src/main/java/ru/yandex/yamblz/ui/activities/MainActivity.java:
```diff
@@ -26,11 +54,124 @@ protected void onCreate(@Nullable Bundle savedInstanceState) {
+                    }
+                    updateArtistsInFragmentsByTag(ViewPagerFragment.TAG, artists);
+                    updateArtistsInFragmentsByTag(ArtistsMasterFragment.TAG, artists);
+                });
```
Нужно отписываться в onStop или где-то ещё.

zagayevskiy commented on app/src/main/java/ru/yandex/yamblz/ui/fragments/ArtistContentFragment.java:
```diff
@@ -0,0 +1,100 @@
+    @Override
+    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
+        super.onViewCreated(view, savedInstanceState);
+        ButterKnife.bind(this, view);
```
В onDestroyView надо делать unbind, чтобы избежать утечки вью.

zagayevskiy commented on app/src/main/java/ru/yandex/yamblz/ui/fragments/ArtistDetailsFragment.java:
```diff
@@ -0,0 +1,73 @@
+        Fragment targetFragment = getTargetFragment();
+        int artistId = getArguments().getInt(TAG_ARTIST_ID);
+
+        if (targetFragment != null && targetFragment instanceof ArtistProviderInterface) {
```
Лучше проверять в onAttach - делаешь это в начале и забываешь. В случае, если интерфейс не реализован, бросай исключение, потому что fail fast. Про != null уже писал.

Tayrinn committed:
fix all comments

fix !=null check, extends all fragments from BaseFragment, do unbind ButterKnife in dialog, remove unused fragment

Tayrinn commented:
Ну как-то отправлять на ПР непроверенный код стрёмно =)
Все комментарии поправила.

zagayevskiy commented on app/src/main/java/ru/yandex/yamblz/ui/fragments/ArtistContentFragment.java:
```diff
@@ -0,0 +1,98 @@
+        if (parent instanceof OnArtistMoreClickListener) {
+            ((OnArtistMoreClickListener)parent).onMoreClick(getArguments().getInt(TAG_ARTIST_ID), imageViewCover);
+        } else {
+            throw new ClassCastException(getString(R.string.not_implemented_on_artist_more_click_listener));
```
Сообщения в исключениях лучше не локализовывать, а писать на английском прямо текстом. Если они вдруг попадут в системы метрик типа Crashlytics, не факт, что такая система сможет правильно их смержить.

Плюс конкретно в таких местах лучше явно указывать имя класса, который должен быть имплементирован - OnArtistMoreClickListener.class.getName()
